### PR TITLE
Refactor `Areas` to `FileLocations`

### PR DESCRIPTION
**Pull request by bobdenotter**

### CODE REVIEW

**Comment by JarJak** [comment]
Closes #396

File: src/Configuration/FileLocation.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        $this->basepath = $basepath;
+        $this->show_all = $show_all;
+    }
+
+    public function getKey(): string
+    {
+        return $this->key;
+    }
+
+    public function setKey(string $key): self
```
Comment (JarJak) [inline src/Configuration/FileLocation.php]: Value Object should be immutable, so no setters

File: src/Configuration/FileLocations.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+declare(strict_types=1);
+
+namespace Bolt\Configuration;
+
+use Tightenco\Collect\Support\Collection;
+
+final class FileLocations
+{
+    /** @var Collection */
+    private $areas;
```
Comment (JarJak) [inline src/Configuration/FileLocations.php]: locations

File: src/Configuration/FileLocations.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Bolt\Configuration;
+
+use Tightenco\Collect\Support\Collection;
+
+final class FileLocations
```
Comment (JarJak) [inline src/Configuration/FileLocations.php]: why final?

File: src/Configuration/FileLocations.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    /** @var Config */
+    private $config;
+
+    /**
+     * Areas constructor.
+     */
+    public function __construct(Config $config)
+    {
+        $this->config = $config;
+        $this->initAreas();
```
Comment (JarJak) [inline src/Configuration/FileLocations.php]: locations

File: src/Configuration/FileLocations.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+final class FileLocations
+{
+    /** @var Collection */
+    private $areas;
+
+    /** @var Config */
+    private $config;
+
+    /**
+     * Areas constructor.
```
Comment (JarJak) [inline src/Configuration/FileLocations.php]: remove comment

File: src/Content/MediaFactory.php [hunk 6]
```diff
@@ -60,7 +60,7 @@ public function createOrUpdateMedia(SplFileInfo $file, string $area, ?string $ti
             $media = new Media();
             $media->setFilename($file->getFilename())
                 ->setPath($path)
-                ->setArea($area);
+                ->setArea($fileLocation);
```
Comment (JarJak) [inline src/Content/MediaFactory.php]: setLocation

File: src/Controller/Backend/Async/UploadController.php [hunk 7]
```diff
@@ -55,10 +55,10 @@ public function handleUpload(Request $request): JsonResponse
             return new JsonResponse(['error' => ['message' => 'Invalid CSRF token']], Response::HTTP_FORBIDDEN);
         }
 
-        $area = $request->query->get('area', '');
+        $location = $request->query->get('location', '');
```
Comment (JarJak) [inline src/Controller/Backend/Async/UploadController.php]: $locationName

File: src/Configuration/FileLocations.php [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public function get(string $area): ?FileLocation
+    {
+        return $this->areas->get($area);
+    }
+
+    private function initAreas(): void
+    {
+        $this->areas = new Collection([
```
Comment (JarJak) [inline src/Configuration/FileLocations.php]: no need for it to be a collection

File: src/Content/MediaFactory.php [hunk 9]
```diff
@@ -33,25 +33,25 @@ class MediaFactory
 
     /** @var Collection */
     private $mediaTypes;
-    private $areas;
+    private $fileLocations;
```
Comment (JarJak) [inline src/Content/MediaFactory.php]: add new line before

File: src/Controller/Backend/Async/UploadController.php [hunk 10]
```diff
@@ -82,11 +82,11 @@ public function handleUpload(Request $request): JsonResponse
         });
 
         /** @var File $result */
-        $result = $uploadHandler->process($request->files->all());
+        $result = $uploadHandler->process($_FILES);
```
Comment (JarJak) [inline src/Controller/Backend/Async/UploadController.php]: undo this change, don't use globals

File: src/Controller/Backend/MediaController.php [hunk 11]
```diff
@@ -22,13 +22,13 @@ class MediaController extends TwigAwareController
     /** @var ObjectManager */
     private $em;
 
-    /** @var Areas */
+    /** @var FileLocations */
     private $areas;
```
Comment (JarJak) [inline src/Controller/Backend/MediaController.php]: $fileLocations

File: src/DataFixtures/ContentFixtures.php [hunk 12]
```diff
@@ -34,10 +34,10 @@ class ContentFixtures extends BaseFixture implements DependentFixtureInterface
     /** @var Config */
     private $config;
 
-    /** @var Areas */
+    /** @var FileLocations */
     private $areas;
```
Comment (JarJak) [inline src/DataFixtures/ContentFixtures.php]: $fileLocations

File: src/DataFixtures/ImagesFixtures.php [hunk 13]
```diff
@@ -23,12 +23,12 @@ class ImagesFixtures extends BaseFixture
     /** @var MediaFactory */
     private $mediaFactory;
 
-    /** @var Areas */
+    /** @var FileLocations */
     private $areas;
```
Comment (JarJak) [inline src/DataFixtures/ImagesFixtures.php]: $fileLocations

File: templates/_partials/fields/geolocation.html.twig [hunk 14]
```diff
@@ -45,7 +45,7 @@
                         <button type="button" class="btn btn-sm btn-default expand justify-content-end"><i class="fa fa-expand"></i></button>
                         <button type="button" class="btn btn-sm btn-default compress"><i class="fa fa-compress "></i></button>
                         <div class="mapholder" style="position: relative; overflow: hidden;">
-                            <a class="editor__image--preview-image" href="/thumbs/kitten.jpg?w=1200&amp;h=1200&amp;area=files&amp;s=1e1ae532b569d01e17644e1c31b12fe9" style="background-image: url(&quot;/thumbs/kitten.jpg?w=1200&amp;h=1200&amp;area=files&amp;s=1e1ae532b569d01e17644e1c31b12fe9&quot;);"></a>
+                            <a class="editor__image--preview-image" href="/thumbs/kitten.jpg?w=1200&amp;h=1200&amp;s=1e1ae532b569d01e17644e1c31b12fe9" style="background-image: url(&quot;/thumbs/kitten.jpg?w=1200&amp;h=1200&amp;s=1e1ae532b569d01e17644e1c31b12fe9&quot;);"></a>
```
Comment (JarJak) [inline templates/_partials/fields/geolocation.html.twig]: use asset or thumbnail

File: src/Configuration/FileLocation.php (hunk 1 above)
Comment (bobdenotter) [reply src/Configuration/FileLocation.php]: Done

File: src/Configuration/FileLocations.php (hunk 3 above)
Comment (bobdenotter) [reply src/Configuration/FileLocations.php]: It was `final` already, just kept it.. I've read it's good practice to do so, and I think i agree: https://matthiasnoback.nl/2018/09/final-classes-by-default-why/

File: src/Configuration/FileLocations.php (hunk 4 above)
Comment (bobdenotter) [reply src/Configuration/FileLocations.php]: done

File: src/Content/MediaFactory.php (hunk 6 above)
Comment (bobdenotter) [reply src/Content/MediaFactory.php]: Done. I've updated the `Media` Entity, so it'll need a `make db-reset`. :-)

File: src/Controller/Backend/Async/UploadController.php (hunk 7 above)
Comment (bobdenotter) [reply src/Controller/Backend/Async/UploadController.php]: It wasn't `areaName` before. I'd like to keep the URL parameters brief.

File: src/Configuration/FileLocations.php (hunk 8 above)
Comment (bobdenotter) [reply src/Configuration/FileLocations.php]: Modified

File: src/Content/MediaFactory.php (hunk 9 above)
Comment (bobdenotter) [reply src/Content/MediaFactory.php]: .. and typehint

File: src/Controller/Backend/Async/UploadController.php (hunk 10 above)
Comment (bobdenotter) [reply src/Controller/Backend/Async/UploadController.php]: I'm not sure we can.. The `process` method in the handler breaks when we pass in SF's files. It really needs the `$_FILES` array, and not an array of `Symfony\Component\HttpFoundation\File\UploadedFile`. The alternative would be to PR a fix to https://github.com/siriusphp/upload.

https://github.com/siriusphp/upload/blob/master/src/Handler.php#L196

screenshot: 

Screenshot 2019-04-14 at 08 08 28

File: src/DataFixtures/ContentFixtures.php (hunk 12 above)
Comment (bobdenotter) [reply src/DataFixtures/ContentFixtures.php]: done

File: src/DataFixtures/ImagesFixtures.php (hunk 13 above)
Comment (bobdenotter) [reply src/DataFixtures/ImagesFixtures.php]: done

File: templates/_partials/fields/geolocation.html.twig (hunk 14 above)
Comment (bobdenotter) [reply templates/_partials/fields/geolocation.html.twig]: This is just a dumb placeholder. See #123 and #118

File: src/Controller/Backend/MediaController.php (hunk 11 above)
Comment (bobdenotter) [reply src/Controller/Backend/MediaController.php]: done

File: src/Configuration/FileLocations.php (hunk 3 above)
Comment (JarJak) [reply src/Configuration/FileLocations.php]: There is a fixer in ECS that makes all classes final and there is a good reason not to use it: you can't mock final classes in tests.

Comment (JarJak) [reply src/Configuration/FileLocations.php]: Also you block user to extend the class, why? Final classes shouldn't be used without interfaces: https://ocramius.github.io/blog/when-to-declare-classes-final/

File: src/Controller/Backend/Async/UploadController.php (hunk 10 above)
Comment (JarJak) [reply src/Controller/Backend/Async/UploadController.php]: `$request->files->all()` does not work?

Comment (JarJak) [reply src/Controller/Backend/Async/UploadController.php]: Why do we use Sirius's upload instead of Symfony mechanisms?

Comment (bobdenotter) [reply src/Controller/Backend/Async/UploadController.php]: `Request->files->all` gives the error in the screenshot. 

> Why do we use Sirius's upload instead of Symfony mechanisms?

Last time i looked into it, handling them yourself in SF is a bit of a chore. Even the page itself suggests to use a third-party module: https://symfony.com/doc/current/controller/upload_file.html

I didn't like that VichUploaderBundle though, and we used this in Bolt 3 as well, and it also works well with async uploads from Vue.

File: src/Configuration/FileLocations.php (hunk 3 above)
Comment (bobdenotter) [reply src/Configuration/FileLocations.php]: Removed. :-)

File: src/Controller/Backend/Async/UploadController.php (hunk 10 above)
Comment (JarJak) [reply src/Controller/Backend/Async/UploadController.php]: I think Symfony Community (me included) is more familiar with VichUploaderBundle than sirius. And because Sirius expects files to be $_FILES I would assume there are more places with bad habits inside their code.

Also it's really not that hard to implement our own uploading mechanism. In the back of head we need to have filesystem abstraction, like Flysystem.

File: src/Controller/Backend/Async/UploadController.php (hunk 7 above)
Comment (JarJak) [reply src/Controller/Backend/Async/UploadController.php]: short url parameter is ok, but variable name can be changed

File: src/Controller/Backend/Async/UploadController.php (hunk 10 above)
Comment (bobdenotter) [reply src/Controller/Backend/Async/UploadController.php]: Let's add a @todo for that, then? Also, latest version of that bundle requires PHP 7.2: https://github.com/dustin10/VichUploaderBundle/blob/1.9.0/composer.json#L23

Comment (JarJak) [reply src/Controller/Backend/Async/UploadController.php]: Yeah, stuff starts requiring this version of PHP... maybe we could break with PHP 7.1?

File: src/Configuration/FileLocation.php [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+namespace Bolt\Configuration;
+
+class FileLocation
+{
+    private $key;
+    private $name;
+    private $basepath;
+    private $show_all = false;
+
+    public function __construct(string $key, string $name, string $basepath, bool $show_all)
```
Comment (JarJak) [inline src/Configuration/FileLocation.php]: $showAll

File: src/Configuration/FileLocation.php [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+declare(strict_types=1);
+
+namespace Bolt\Configuration;
+
+class FileLocation
+{
+    private $key;
+    private $name;
+    private $basepath;
+    private $show_all = false;
```
Comment (JarJak) [inline src/Configuration/FileLocation.php]: $showAll

File: src/DataFixtures/ContentFixtures.php (hunk 12 above)
Comment (JarJak) [reply src/DataFixtures/ContentFixtures.php]: still not done

Comment (bobdenotter) [reply src/DataFixtures/ContentFixtures.php]: weird, i could've sworn i did.. It'll e in the next `push`!

File: src/Configuration/FileLocation.php (hunk 15 above)
Comment (bobdenotter) [reply src/Configuration/FileLocation.php]: done

File: src/Configuration/FileLocation.php (hunk 16 above)
Comment (bobdenotter) [reply src/Configuration/FileLocation.php]: done

File: src/Controller/Backend/Async/UploadController.php (hunk 10 above)
Comment (bobdenotter) [reply src/Controller/Backend/Async/UploadController.php]: I am very adamant about that: We should stick with SF's requirements. We are not Marco! ;-)
